File: src/Main/lua/ga/corelauncher/Classes/GameManager.lua
local GameManager = {}
GameManager.Games = {}

local Base64Img = TypeWriter.JavaScript:Require("base64-img")
local FS = TypeWriter.JavaScript:Require("fs-extra")

local GameManager = Import("ga.corelauncher.Libraries.ClassCreator")(
    "GameManager",
    function(self)

    end,
    {
        LoadGames = function(self, Games)
            self.Games = Games
        end,
        ListGames = function(self)
            return self.Games
        end,
        GetGame = function(self, GameId)
            return self:ListGames()[GameId]
        end,

        --#region Icon getting
        GetGameIcon = function(self, GameId)
            local IconPath = TypeWriter.ResourceManager:GetFilePath(self:GetGame(GameId).Icon)
            return FS:readFileSync(IconPath, "utf8")
        end,
        GetGameIconBase64 = function(self, GameId)
            local IconPath = TypeWriter.ResourceManager:GetFilePath(self:GetGame(GameId).Icon)
            return Base64Img:base64Sync(IconPath)
        end,
        --#endregion

        --#region Banner getting
        GetGameBanner = function(self, GameId)
            local BannerPath = TypeWriter.ResourceManager:GetFilePath(self:GetGame(GameId).Banner)
            return FS:readFileSync(BannerPath, "utf8")
        end,
        GetGameBannerBase64 = function(self, GameId)
            local BannerPath = TypeWriter.ResourceManager:GetFilePath(self:GetGame(GameId).Banner)
            return Base64Img:base64Sync(BannerPath)
        end,
        --#endregion

        GetValidAccounts = function(self, GameId)
            local Accounts = CoreLauncher.AccountManager:ListAccounts()
            local ValidAccounts = {}
            for _, Account in ipairs(Accounts) do
                local IsAccountValid = self:GetGame(GameId).IsAccountValid(Account)
                if IsAccountValid then
                    table.insert(ValidAccounts, Account)
                end
            end
            return ValidAccounts
        end,

        ListInstanceVersions = function(self, GameId)
            return self:GetGame(GameId).InstanceVersions
        end,

        ListInstanceVersionValues = function(self, GameId, InstanceVersionId, PriorValues)
            local InstanceVersions = self:GetGame(GameId).InstanceVersions
            local InstanceVersion
            for _, Version in pairs(InstanceVersions) do
                if Version.Id == InstanceVersionId then
                    InstanceVersion = Version
                    break
                end
            end
            for Key, Value in pairs(PriorValues) do
                if Value == "undefined" then
                    PriorValues[Key] = nil
                end
            end
            local ObtainedValues = InstanceVersion.ObtainValues(PriorValues)
            Inspect(ObtainedValues)
            return ObtainedValues
        end
    }
)

return GameManager

File: src/Main/js/ga/corelauncher/Classes/GameManager.js
const ResourceBase64 = Import("ga.corelauncher.Libraries.ResourceBase64")
const UUID = require("uuid").v4

class GameManager {
    constructor(Games) {
        this.Games = Games

        for (const GameId in Games) {
            CoreLauncher.DataBase.SetKeyIfNotExists(`Game.${GameId}.Properties`, this.ListDefaultGameProperties(GameId))
            CoreLauncher.DataBase.SetKeyIfNotExists(`Game.${GameId}.Instances`, {})

        }
    }

    ListGames() {
        return this.Games
    }

    GetGame(GameId) {
        return this.Games[GameId]
    }

    //#region Images
    GetGameIconBase64(GameId) {
        return ResourceBase64(this.Games[GameId].Icon)
    }

    GetGameBannerBase64(GameId) {
        return ResourceBase64(this.Games[GameId].Banner)
    }

    GetValidAccounts(GameId) {
        const Accounts = CoreLauncher.AccountManager.ListAccounts()
        const ValidAccounts = Accounts.filter(this.Games[GameId].IsAccountValid)
        return ValidAccounts
    }
    //#endregion

    //#region Game properties
    ListGameProperties(GameId) {
        return this.Games[GameId].GameProperties
    }

    ListDefaultGameProperties(GameId) {
        const Properties = this.ListGameProperties(GameId)
        const DefaultProperties = {}

        for (const PropertyRow of Properties) {
            for (const Property of PropertyRow) {
                DefaultProperties[Property.Id] = Property.Default
            }
        }

        return DefaultProperties
    }

    SetGameProperties(GameId, Properties) {
        CoreLauncher.DataBase.SetKey(`Game.${GameId}.Properties`, Properties)
    }

    GetGameProperties(GameId) {
        return CoreLauncher.DataBase.GetKey(`Game.${GameId}.Properties`)
    }
    //#endregion

    ListInstanceVersions(GameId) {
        return this.Games[GameId].InstanceVersions
    }

    async ListInstanceVersionValues(GameId, InstanceVersionId, Values) {
        const InstanceVersions = this.ListInstanceVersions(GameId)
        const InstanceVersion = InstanceVersions.filter(InstanceVersion => InstanceVersion.Id == InstanceVersionId)[0]
        const ObtainedValues = await InstanceVersion.ObtainValues(Values)
        return ObtainedValues
    }

    ListInstanceProperties(GameId, Rows = true) {
        const Properties = this.Games[GameId].InstanceProperties
        if (Rows) { return Properties }
        return Properties.flat()
    }

    CreateInstance(GameId, InstanceData) {
        InstanceData.UUID = UUID()
        CoreLauncher.DataBase.SetKey(`Game.${GameId}.Instances.${InstanceData.UUID}`, InstanceData)
    }

    RemoveInstance(GameId, InstanceId) {
        CoreLauncher.DataBase.RemoveKey(`Game.${GameId}.Instances.${InstanceId}`)
    }

    ListInstances(GameId) {
        return Object.values(CoreLauncher.DataBase.GetKey(`Game.${GameId}.Instances`))
    }

    GetInstance(GameId, InstanceId) {
        return CoreLauncher.DataBase.GetKey(`Game.${GameId}.Instances.${InstanceId}`)
    }

    GetInstanceProperties(GameId, InstanceId, FillDefaults = true) {
        const Properties = this.GetInstance(GameId, InstanceId).Properties
        if (!FillDefaults) {
            return Properties
        }
        var ReturnProperties = {}
        const DefaultProperties = this.ListInstanceProperties(GameId, false)

        for (const Property of DefaultProperties) {
            console.log(Property)
            if (Properties[Property.Id] == null) {
                ReturnProperties[Property.Id] = Property.Default
            } else {
                ReturnProperties[Property.Id] = Properties[Property.Id]
            }
        }

        return this.GetInstance(GameId, InstanceId).Properties
    }

    SetInstanceProperties(GameId, InstanceId, Properties) {
        CoreLauncher.DataBase.SetKey(`Game.${GameId}.Instances.${InstanceId}.Properties`, Properties)
    }

    

}

module.exports = GameManager